subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -286,28 +285,25 @@ public void apply(final Project project) {
         BuildOutputCleanupRegistry buildOutputCleanupRegistry = projectInternal.getServices().get(BuildOutputCleanupRegistry.class);
 
         SourceSet mainSourceSet = javaExtension.getSourceSets().create(SourceSet.MAIN_SOURCE_SET_NAME);
-        configureSourceSets(project, javaExtension, buildOutputCleanupRegistry, mainSourceSet);
-        configureConfigurations(project, javaExtension, mainSourceSet);
+        PublishArtifact jarArtifact = configureArchives(project, mainSourceSet);
 
-        configureJavadocTask(project, javaExtension);
-        configureArchivesAndComponent(project, javaExtension);
+        configureBuiltInTest(project, mainSourceSet);
+        configureSourceSets(javaExtension, buildOutputCleanupRegistry);
+        configureConfigurations(project, mainSourceSet, jarArtifact);
+        configureJavaDocTask(null, mainSourceSet, project.getTasks(), javaExtension);
         configureBuild(project);
     }
 
-    private void configureSourceSets(Project project, JavaPluginExtension pluginExtension, final BuildOutputCleanupRegistry buildOutputCleanupRegistry, SourceSet mainSourceSet) {
-        SourceSetContainer sourceSets = pluginExtension.getSourceSets();
-
-        // The built-in test suite must be configured after the main source set is available due to some
-        // special handling in the IntelliJ model builder
-        configureBuiltInTest(project, project.getExtensions().getByType(TestingExtension.class), pluginExtension, mainSourceSet);
-
+    private static void configureSourceSets(JavaPluginExtension pluginExtension, final BuildOutputCleanupRegistry buildOutputCleanupRegistry) {
         // Register the project's source set output directories
-        sourceSets.all(sourceSet ->
-            buildOutputCleanupRegistry.registerOutputs(sourceSet.getOutput())
-        );
+        pluginExtension.getSourceSets().all(sourceSet -> buildOutputCleanupRegistry.registerOutputs(sourceSet.getOutput()));
     }
 
-    private void configureBuiltInTest(Project project, TestingExtension testing, JavaPluginExtension java, SourceSet mainSourceSet) {
+    /**
+     * Must be called after the main source set is available due to some special handling in the IntelliJ model builder.
+     */
```

Comment:
Is this comment still relevant?  Within the scope of this method, the concept of a main SourceSet doesn't seem to matter much, it's purpose is to setup the default test suite using the given SourceSet argument.  The SourceSet must be available, since we're calling methods on it (after the suite is configured, which is forced by the `get()` on line 316).

Is it still worth keeping this comment?  And if so, with the prominence of the only line of javadoc on this method vs. a line comment within the method body?

Reply:
It seems the previous implementation was configuring the built-in tests as part of test-suites-plugin, but was using a callable referencing the main source set when configuring the tests's compile/runtime classpath. Apparently intelliJ cares enough about the ordering such that we can't lazily reference the main source set. So, we need to make sure we define the test suite after the main source set is defined. See: https://github.com/gradle/gradle/pull/18561/files

I'll update the comment with a little more details and move it out of the javadoc